### PR TITLE
run_tests.yml: test against multiple pandoc versions

### PR DESCRIPTION
**Pull request by ickc**
This PR add pypy3 test, and also test against multiple pandoc versions.

May not be ready to merged if discussion of supported pandoc versions is needed. By the way, is it mentioned somewhere in the doc which pandoc versions are supported?

By the way, the run test and publish workflow can be combined into one. E.g. setup in <https://github.com/ickc/pantable/blob/master/.github/workflows/pythonpackage.yml>.

Edit: seems like pandoc v1 support is dropped as shown in <https://github.com/ickc/panflute/runs/676485536>.

Edit2: pypy3 is added because it is stated as a supporting platform in `setup.py`.

Edit3: from [CI](https://github.com/ickc/panflute/actions/runs/105202745) seems like only down to pandoc 2.7 is supported.

### CODE REVIEW

File: .github/workflows/run_tests.yml [hunk 1]
```diff
@@ -21,7 +21,9 @@ jobs:
       fail-fast: false
       max-parallel: 5
       matrix:
-        python-version: [3.6, 3.7, 3.8] # , pypy3]
+        python-version: [3.6, 3.7, 3.8, pypy3]
```
Comment (dhimmel) [inline .github/workflows/run_tests.yml:24]: 3.9 is now available. Probably more important to test different pandoc versions than python versions, although testing the python extremes (3.6 and 3.9) is definitely helpful

File: .github/workflows/run_tests.yml [hunk 2]
```diff
@@ -42,12 +44,10 @@ jobs:
...
     - name: Download Pandoc
       run: |
-        url="https://github.com/jgm/pandoc/releases/latest"
-        path=$(curl -L $url | grep -o '/jgm/pandoc/releases/download/.*-amd64\.deb')
-        downloadUrl="https://github.com$path"
-        file=${path##*/}
-        wget $downloadUrl &&
-        sudo dpkg -i $file
+        pip install requests
+        downloadUrl=$(python -c 'import requests, sys; v = sys.argv[1]; url = "https://api.github.com/repos/jgm/pandoc/releases/" + ("" if v == "latest" else "tags/") + v; print(next(i["browser_download_url"] for i in requests.get(url).json()["assets"] if i["name"][-3:] == "deb"))' ${{ matrix.pandoc-version }})
```
Comment (dhimmel) [inline .github/workflows/run_tests.yml:48]: almost feel that this is so complex is should be its own script that is called? Rather than a hard-to-read line.

**Review by dhimmel — COMMENTED** [review]
would be helpful for testing issues like https://github.com/sergiocorreia/panflute/issues/142

File: .github/workflows/run_tests.yml (hunk 2 above)
Comment (ickc) [reply .github/workflows/run_tests.yml:48]: The before and after scripts are both written by me. The former regex the html directly, the later uses the GitHub API itself. While it might have readability issue, it is rarely the thing you need to debug as it is stable. (I've been using both in various projects for years.)

I wouldn't worry about this too much. The main point is to reintroduce multi Python and pandoc versions for testing. I think when I first set up Travis CI for panflute it has much more combinations.

File: .github/workflows/run_tests.yml (hunk 1 above)
Comment (ickc) [reply .github/workflows/run_tests.yml:24]: Unfortunately this is a Python project. It seems that the standard is to test against all supported Python versions (as declared in setup.py.) Pandoc versions on the other hand, can be tested sparingly from experience, because the API is stable. Often you only need to test against "breaking boundaries", such as 1.17(?), 1.x, 2.0.x, 2.9.x, 2.11, latest. In my repos I sometimes toggle all pandoc versions once in a long while to make sure the gaps has no problem.

Comment (ickc) [reply .github/workflows/run_tests.yml:24]: also see commit 201815b and the comment above about pandoc versions supported.

when the PR is created Python 3.9 is not released yet.

**Comment by ickc** [comment]
Updated for more Python and pandoc versions. Note that the only failing ones are pandoc >=2.10 which is not supported yet. So this PR is passing per se.

Also note that GitHub Action doesn't support "allow failures", c.f. notes in the code and https://github.com/actions/toolkit/issues/399

**Comment by alerque** [comment]
> Also note that GitHub Action doesn't support "allow failures", c.f. notes in the code and [actions/toolkit#399](https://github.com/actions/toolkit/issues/399)

It does support marking certain tests from a matrix as "required" and others not. The setup for this is not in the workflow setup file it is in the repository settings under protected branches.

**Comment by ickc** [comment]
Feel free to PR in that case.

**Comment by ickc** [comment]
> GitHub Actions does support marking certain tests from the job matrix as "required" and others not. The setup for this is not in the workflow config file it is in the repository settings under protected branches.

Do you mean the "Require status checks to pass before merging" in "Branch protection rule"? That would be helpful.

**Comment by sergiocorreia** [comment]
Open question: we need a policy on what versions of panflute to support.

With Python, it's quite easy to just support whatever is not EOL (in this case 3.6-3.9):

1. There are not many versions
2. They don't really change much

The only thing we might be missing by supporting 3.6 is fstrings, so if not many people use 3.6 then we can ditch it (it's EOL at the end of next year anyways).

However, with pandoc it's harder:

1. There is no clear concept of EOL (which is understandable!)
2. Whenever the API changes, it would mean we have to keep multiple versions of code to support old-table, new-table, etc.

Here, my inclination would be to only support `latest-pandoc`, but keep a table in the readme indicating, for each `pandoc x.y` version, the last version of panflute that supports it. Thus, anyone that has an old pandoc can just use an earlier version of panflute, and we don't have to waste valuable time supporting multiple pandoc versions.

**Comment by ickc** [comment]
I think the testing versions and supported versions can be decided independently. As @alerque suggested, it can be done using "Require status checks to pass before merging" in "Branch protection rule". I think testing against a few pandoc versions is useful to watch out for potential problems, but should not declare support on all that's tested. In fact, similar to `allow_failure` in Travis, and the branch protection rule in GitHub Action, some are allowed to fail.

e.g. this is the reason why pandoc also test against the latest HEAD of GHC but allow them to fail.

By my test, currently panflute is working well with pandoc 2.7–2.9. That information is useful.

Also, I think there should be no promise on what versions is supported. Say if implementing support for 2.11 breaks those in 2.7-2.9, then drop it. i.e. the causality is opposite: knowing which is passing inform you which to suport, but not vice versa.

**Comment by dhimmel** [comment]
> With Python, it's quite easy to just support whatever is not EOL (in this case 3.6-3.9):

Agree. Also f-strings arrived in 3.6.

> Here, my inclination would be to only support latest-pandoc, but keep a table in the readme indicating, for each pandoc x.y version, the last version of panflute that supports it.

Ubuntu 14.04 LTS reaches EOL on April 2022, and [ships with pandoc 1.12.2.1](https://launchpad.net/ubuntu/trusty/+package/pandoc). I notice a lot of cases where users are stuck with old Pandoc versions, since they install whatever version is available from their system's package manager. So I reason:

1. there is a large benefit to users to maintaining compatibility with old Pandoc versions
2. there is sometimes, but not always, a large development cost to back compat

So I'd propose continue supporting old pandoc versions while CI indicates panflute will work on them. But when an AST change occurs and supporting old Pandoc versions is too burdensome, drop the old support.

One final thing is that it'd be great to have panflute's support of Pandoc's versions programmed into the codebase. One solution would be for the code to log a warning to stderr should Panflute receive an AST version that it does not support. Or you could even have it fail by default, with a metadata value like `panflute-ignore-unsupported-AST: true` to override the failure.

This is great discussion! But getting a bit further away from this PR. The PR looks acceptable to me, once the failing Pandoc versions are marked as non-required or dropped.

**Review by sergiocorreia — APPROVED** [review]
Merging it, but will probably need to change it once we are done adding support for pandoc 2.10, as that will probably break support for tables in older pandoc versions. We can still try to run older pandoc versions, but need to allow failure
